### ISSUE TITLE
Problem Training Network for Image segmentation using u-net like network [MXNET & R]

### ISSUE THREAD

**Issue by Jorisfournel** [dmlc/mxnet#5003]
I'm trying to build a modest network for image segmentation in medical images for a school project,
I'm a total beginner by the way at this.
So I tried to build a network that is structured like the u-net, here is my code,
when i run with the gpu device, rstudio just end up by crashing without error message, although I get one when I make it run from the console in R, the images I'm using here to train are from the ISBI Challenge 2012 (neuronal images) (http://brainiac2.mit.edu/isbi_challenge/home)

I would like to understand why it is crashing because I don't have a clue for now, there is probably something to set more correctly and precisely for the output layer (softmax) since this a segmentation problem and each pixel has a true label, also i got some issues with the dimension of the train.x and train.y, but it seemed to be solved ...

When I run with the cpu and just 1 training epoch, i get a model ...

For 2 epochs and more i get this error message : 

"Error in exec$update.arg.arrays(arg.arrays, match.name, skip.null) : 
  executor.cc:65: RCheck failed: to->containsElementNamed(names[i].c_str()) cannot find key NA in the array arg.arrays"

MXNet version: 9.4

If you are using R package, please provide

R `sessionInfo()`: R version 3.3.2 (2016-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.1 LTS

locale:
 [1] LC_CTYPE=fr_FR.UTF-8       LC_NUMERIC=C               LC_TIME=fr_FR.UTF-8       
 [4] LC_COLLATE=fr_FR.UTF-8     LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=fr_FR.UTF-8   
 [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] jpeg_0.1-8  mxnet_0.9.4

loaded via a namespace (and not attached):
 [1] igraph_1.0.1       Rcpp_0.12.9        rstudioapi_0.6     magrittr_1.5      
 [5] munsell_0.4.3      colorspace_1.3-2   R6_2.2.0           brew_1.0-6        
 [9] stringr_1.1.0      plyr_1.8.4         dplyr_0.5.0        visNetwork_1.0.3  
[13] Rook_1.1-1         tools_3.3.2        grid_3.3.2         gtable_0.2.0      
[17] DBI_0.5-1          influenceR_0.1.0   DiagrammeR_0.9.0   htmltools_0.3.5   
[21] lazyeval_0.2.0     digest_0.6.12      assertthat_0.1     tibble_1.2        
[25] gridExtra_2.2.1    RColorBrewer_1.1-2 ggplot2_2.2.1      codetools_0.2-15  
[29] htmlwidgets_0.8    viridis_0.3.4      rgexf_0.15.3       stringi_1.1.2     
[33] scales_0.4.1       XML_3.98-1.5       jsonlite_1.2      

## Error Message:
1.Rstudio crashes with GPU
2."Error in exec$update.arg.arrays(arg.arrays, match.name, skip.null) : 
  executor.cc:65: RCheck failed: to->containsElementNamed(names[i].c_str()) cannot find key NA in the array arg.arrays" with CPU

**MY CODE :**

```r
require(mxnet)

require(jpeg)
#Echantillon train

files <- list.files(path="/home/fournel/Documents/Projet/Neurones set/train-volume")
a='/home/fournel/Documents/Projet/Neurones set/train-volume/'                        
filess=paste(a,files, sep='')
list_of_images = lapply(filess, readJPEG) 

train.x = do.call('cbind', lapply(list_of_images, as.vector))
train.array <- train.x
dim(train.array) <- c(512,512,1,30)



#Labels

files <- list.files(path="/home/fournel/Documents/Projet/Neurones set/train-labels")
b='/home/fournel/Documents/Projet/Neurones set/train-labels/'                        
filess=paste(b,files, sep='')
list_of_images = lapply(filess, readJPEG) 
train.y = do.call('cbind', lapply(list_of_images, as.vector))

train.y[which(train.y<0.5)]=0
train.y[which(train.y>0.5)]=1
train.y.array=train.y
dim(train.y.array)=c(512,512,1,30)


data <- mx.symbol.Variable("data")
d0b <- mx.symbol.Convolution(data, name="d0b",kernel=c(3,3), num.filter=64)
act1 <- mx.symbol.Activation(d0b, name="relu1", act_type="relu")
d0c <- mx.symbol.Convolution(act1, name="d0c", num_filter=64, kernel=c(3,3))
act2 <- mx.symbol.Activation(d0c, name="relu2", act_type="relu")

d1a <- mx.symbol.Pooling(act2, name='pool_d0c-1a', kernel=c(2,2), stride=c(2,2), pool_type='max')
d1b <- mx.symbol.Convolution(d1a, name='conv_d1a-b', kernel=c(3,3), num.filter=128)
actd1b <- mx.symbol.Activation(d1b, name='relu_d1b', act_type="relu")
d1c <- mx.symbol.Convolution(actd1b, name='conv_d1b-c', kernel=c(3,3), num.filter=128)
actd1c <- mx.symbol.Activation(d1c, name='relu_d1c', act_type='relu')

d2a <- mx.symbol.Pooling(actd1c, name='pool_d1c-2a', kernel=c(2,2), stride=c(2,2), pool_type='max')
d2b <- mx.symbol.Convolution(d2a, name='conv_d2a-b', kernel=c(3,3), num.filter=256)
actd2b <- mx.symbol.Activation(d2b, name='relu_d2b', act_type="relu")
d2c <- mx.symbol.Convolution(actd2b, name='conv_d2b-c', kernel=c(3,3), num.filter=256)
actd2c <- mx.symbol.Activation(d2c, name='relu_d2c', act_type="relu")

d3a <- mx.symbol.Pooling(actd2c, name='pool_d2c-3a', kernel=c(2,2), stride=c(2,2), pool_type='max')
d3b <- mx.symbol.Convolution(d3a, name='conv_d3a-b', kernel=c(3,3), num.filter=512)
actd3b <- mx.symbol.Activation(d3b, name='relu_d3b', act_type="relu")
d3c <- mx.symbol.Convolution(actd3b, name='conv_d3b-c', kernel=c(3,3), num.filter=512)
actd3c <- mx.symbol.Activation(d3c, name='relu_d3c', act_type="relu")
drop1 <- mx.symbol.Dropout(actd3c, name='dropout_d3c')

d4a <- mx.symbol.Pooling(drop1, name='pool_d3c-4a', kernel=c(2,2), stride=c(2,2), pool_type='max')
d4b <- mx.symbol.Convolution(d4a, name='conv_d4a-b', kernel=c(3,3), num.filter=1024)
actd4b <- mx.symbol.Activation(d4b, name='relu_d4b', act_type="relu")
d4c <- mx.symbol.Convolution(actd4b, name='conv_d4b-c', kernel=c(3,3), num.filter=1024)
actd4c <- mx.symbol.Activation(d4c, name='relu_d4c', act_type="relu")
drop2 <- mx.symbol.Dropout(actd4c, name='dropout_d4c')

u3a <- mx.symbol.Deconvolution(drop2, name='upconv_d4c_u3a', kernel=c(2,2), stride=c(2,2), num.filter=512)
actu3a <- mx.symbol.Activation(u3a, name='relu_u3a', act_type="relu")
d3cc <- mx.symbol.Crop(d3c, actu3a,num.args=2, name='crop_d3c-d3cc')
u3b <- mx.symbol.Concat(list(actu3a, d3cc), num.args=2, name='concat_d3cc_u3a-b')
u3c <- mx.symbol.Convolution(u3b, name='conv_u3b-c', kernel=c(3,3), num.filter=512)
actu3c <- mx.symbol.Activation(u3c, name='relu_u3c', act_type="relu")
u3d <- mx.symbol.Convolution(actu3c, name='conv_u3c-d', kernel=c(3,3), num.filter=512)
actu3d <- mx.symbol.Activation(u3d, name='relu_u3d', act_type="relu")

u2a <- mx.symbol.Deconvolution(actu3d, name='upconv_u3d_u2a', kernel=c(2,2), stride=c(2,2), num.filter=256)
actu2a <- mx.symbol.Activation(u2a, name='relu_u2a', act_type="relu")
d2cc <- mx.symbol.Crop(d2c, actu2a,num.args=2, name='crop_d2c-d2cc')
u2b <- mx.symbol.Concat(data=list(actu2a, d2cc),num.args=2, name='concat_d2cc_u2a-b')
u2c <- mx.symbol.Convolution(u2b, name='conv_u2b-c', kernel=c(3,3), num.filter=256)
actu2c <- mx.symbol.Activation(u2c, name='relu_u2c', act_type="relu")
u2d <- mx.symbol.Convolution(actu2c, name='conv_u2c-d', kernel=c(3,3), num.filter=256)
actu2d <- mx.symbol.Activation(u2d, name='relu_u2d', act_type="relu")

u1a <- mx.symbol.Deconvolution(actu2d, name='upconv_u2d_u1a', kernel=c(2,2), stride=c(2,2), num.filter=128)
actu1a <- mx.symbol.Activation(u1a, name='relu_u1a', act_type="relu")
d1cc <- mx.symbol.Crop(d1c, actu1a,num.args=2, name='crop_d1c-d1cc')
u1b <- mx.symbol.Concat(data=list(actu1a, d1cc), num.args=2, name='concat_d1cc_u1a-b')
u1c <- mx.symbol.Convolution(u1b, name='conv_u1b-c', kernel=c(3,3), num.filter=128)
actu1c <- mx.symbol.Activation(u1c, name='relu_u1c', act_type="relu")
u1d <- mx.symbol.Convolution(actu1c, name='conv_u1c-d', kernel=c(3,3), num.filter=128)
actu1d <- mx.symbol.Activation(u1d, name='relu_u1d', act_type="relu")

u0a <- mx.symbol.Deconvolution(actu1d, name='upconv_u1d_u0a', kernel=c(2,2), stride=c(2,2), num.filter=64)
actu0a <- mx.symbol.Activation(u0a, name='relu_u0a', act_type="relu")
d0cc <- mx.symbol.Crop(d0c, actu0a, num.args=2, name='crop_d0c-d0cc')
u0b <- mx.symbol.Concat(data=list(actu0a, d0cc), num.args=2, name='concat_d0cc_u0a-b')
u0c <- mx.symbol.Convolution(u0b, name='conv_u0b-c', kernel=c(3,3), num.filter=64)
actu0c <- mx.symbol.Activation(u0c, name='relu_u0c', act_type="relu")
u0d <- mx.symbol.Convolution(actu0c, name='conv_u0c-d', kernel=c(3,3), num.filter=64)
actu0d <- mx.symbol.Activation(u0d, name='relu_u0d', act_type="relu")

score <- mx.symbol.Convolution(actu0d, name='conv_u0d-score', kernel=c(1,1), num.filter=2)
loss <- mx.symbol.SoftmaxActivation(data=score, name='loss')


devices <- mx.gpu()
mx.set.seed(0)

model <- mx.model.FeedForward.create(loss, X=train.array, y=train.y.array,
                                     ctx=devices, num.round=1,initializer=mx.init.normal(sqrt(2/576)),
                                     learning.rate=0.05, momentum=0.99 )

```

**Comment by thirdwing** [comment]
I think this will also be fixed in #6673

**Comment by thirdwing** [comment]
@Jorisfournel I am afraid there is something wrong in your network configuration.

The output size is not the same as input size.

You can use a helper function to see the output shape of different layers in the network.

```r
print_inferred_shape <- function(net) {
  slist <- mx.symbol.infer.shape(symbol = net, data = c(168, 168, 1, 2))
  print(slist$out.shapes)
}
```
